net/tcp/tcp_recvwindow.c

```diff
@@ -181,6 +181,29 @@ uint16_t tcp_get_recvwindow(FAR struct net_driver_s *dev,
       recvwndo = tailroom;
     }
 
+#if CONFIG_NET_RECV_BUFSIZE > 0
+  /* Finally, check the receive buffer size further to avoid receive
+   * window out of desire.
+   */
```

Comment:
i guess than conn->readahead == NULL case should be treated as io_pktlen == 0.

that is,
```
queued_bytes = FIONREAD equivalent;
sbleft = max(conn->rcv_bufs - queued_bytes, 0);
recvwndo = min(recvwndo, sbleft);
```